### DISCUSSION TITLE
Correctness of ComputeLagrangianWrappedStress

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#21684] [Q&A Modules: Solid mechanics] [answered]
Hello,

I recently experimented with `ComputeLagrangianWrappedStress` and expected that I'd be able to replicate the results I get when using the `StressDivergenceTensors`-System exactly. But that wasn't the case. I used the standard material model of `ComputeFiniteStrain` and `ComputeFiniteStrainElasticStress`. The input is as follows

```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 3
  ny = 3
  nz = 3
  elem_type = HEX8
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    add_variables = true
    generate_output = 'stress_xx stress_yy stress_zz stress_xy'
  []
[]

[Functions]
  [tdisp]
    type = ParsedFunction
    value = '0.5 * t' 
  []
  [tdisp_quer]
    type = ParsedFunction
    value = '0.5 * y * t' 
  []
[]

[BCs]
  [bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [bottom_x]
    type = DirichletBC
    variable = disp_x
    boundary = bottom
    value = 0
  []
  [bottom_z]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0
  []
  [left_x]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [right_x]
    type = DirichletBC
    variable = disp_x
    boundary = right
    value = 0
  []
  [front_z]
    type = DirichletBC
    variable = disp_z
    boundary = front
    value = 0
  []
  [back_z]
    type = DirichletBC
    variable = disp_z
    boundary = back
    value = 0
  []
  [tdisp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = tdisp
  []
[]

[Materials]
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 30
    poissons_ratio = 0.4
  []
  [stress]
    type = ComputeFiniteStrainElasticStress
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type'
    petsc_options_value = 'lu'
  []
[]

[Executioner]
  type = Transient
  solve_type = Newton
  end_time = 1
  dt = 0.25
[]

[Outputs]
  exodus = true
  csv = true
[]

[Postprocessors]
  [cauchy_xx]
    type = ElementAverageValue
    variable = stress_xx
    execute_on = 'initial timestep_end'
  []
  [cauchy_yy]
    type = ElementAverageValue
    variable = stress_yy
    execute_on = 'initial timestep_end'
  []
  [cauchy_xy]
    type = ElementAverageValue
    variable = stress_xy
    execute_on = 'initial timestep_end'
  []
  [cauchy_zz]
    type = ElementAverageValue
    variable = stress_zz
    execute_on = 'initial timestep_end'
  []
[]
```

The corresponding input with lagrangian kernel system is

```
[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 3
  ny = 3
  nz = 3
  elem_type = HEX8
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  #large_kinematics = true
[]

[Modules/TensorMechanics/Master]
  [all]
    new_system = true
    formulation = TOTAL
    strain = FINITE
    add_variables = true
    generate_output = 'cauchy_stress_xx cauchy_stress_yy cauchy_stress_zz cauchy_stress_xy'
  []
[]

[Functions]
  [tdisp]
    type = ParsedFunction
    value = '0.5 * t' 
  []
  [tdisp_quer]
    type = ParsedFunction
    value = '0.5 * y * t' 
  []
[]

[BCs]
  [bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [bottom_x]
    type = DirichletBC
    variable = disp_x
    boundary = bottom
    value = 0
  []
  [bottom_z]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0
  []
  [left_x]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [right_x]
    type = DirichletBC
    variable = disp_x
    boundary = right
    value = 0
  []
  [front_z]
    type = DirichletBC
    variable = disp_z
    boundary = front
    value = 0
  []
  [back_z]
    type = DirichletBC
    variable = disp_z
    boundary = back
    value = 0
  []
  [tdisp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = tdisp
  []
[]

[Materials]
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 30
    poissons_ratio = 0.4
  []
  [stress]
    type = ComputeLagrangianWrappedStress
  []
  [stress_base]
    type = ComputeFiniteStrainElasticStress
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type'
    petsc_options_value = 'lu'
  []
[]

[Executioner]
  type = Transient
  solve_type = Newton
  end_time = 1
  dt = 0.25
[]

[Outputs]
  exodus = true
  csv = true
[]

[Postprocessors]
  [cauchy_xx]
    type = ElementAverageValue
    variable = cauchy_stress_xx
    execute_on = 'initial timestep_end'
  []
  [cauchy_yy]
    type = ElementAverageValue
    variable = cauchy_stress_yy
    execute_on = 'initial timestep_end'
  []
  [cauchy_xy]
    type = ElementAverageValue
    variable = cauchy_stress_xy
    execute_on = 'initial timestep_end'
  []
  [cauchy_zz]
    type = ElementAverageValue
    variable = cauchy_stress_zz
    execute_on = 'initial timestep_end'
  []
[]
```

Not using `large_kinematics` in the `GlobalParams` yields results that are closer to the previous input but still not quite equal. Is the behaviour of not being able to replicate results exactly expected?

Furthermore I noticed that both inputs seem to react to the timestep size. Changing the timestep size changes the stresses in the output. I feel like that shouldn't be the case. What am I missing?

**Comment by reverendbedford** [comment, marked as answer]
1. When using the `ComputeLagrangianWrappedStress` class you should always use a small deformation MOOSE material model.  So in your case you should use `ComputeLinearElasticStress`.  The `ComputeLagrangianWrappedStress` "takes care" of making the small deformation model respond in a large deformations context by advecting the response with an objective stress rate.
2. You should always have the stress calculator use `large_kinematics` when you're using finite strain, as the Jacobian won't be correct otherwise.
3. Even with these changes you won't get an exact match between the old and new models, as we are using a [different](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwibjoOi0Zb5AhX1FlkFHfgBBMMQFnoECCEQAQ&url=https%3A%2F%2Fmooseframework.inl.gov%2Fsource%2Fmaterials%2Flagrangian%2FComputeLagrangianObjectiveStress.html&usg=AOvVaw0JH6sqI5dM45PWwL6QflDN) objective rate.
4. There is a PR to add the [Green-Naghdi](https://github.com/idaholab/moose/pull/21167) objective rate.  If you use that rate (instead of the default Truesdell rate) and take small time steps you should closely match the old formulation.

**Reply by Flolaffel** [reply]
Thanks for your reply. Right now I can say:
1. When using `ComputeLinearElasticStress` instead of `ComputeFiniteStrainElasticStress` in my second input I just get an error that tells me to use  `ComputeFiniteStrainElasticStress` instead. What else do I have to do?
2. Ok, that's what I initially thought too.
3. and 4. I will read into that, thank you.

**Reply by reverendbedford** [reply]
For the second input file?

To be clear: "Lagrangian" kernels need small deformation (ComputeLinearElasticStress) + ComputeLagrangianWrappedStress.  Old kernels need  ComputeFiniteStrainElasticStress.

**Reply by reverendbedford** [reply]
Nevermind this seems to be a bug.  I'll fix it.

**Reply by reverendbedford** [reply]
I partly lied: you *can* use `ComputeFiniteStrainElasticStress` because we zero out the rotations being fed into the original moose material update.  However you should be able to use ComputeLinearElasticStress both practically and logically.

**Reply by reverendbedford** [reply]
Okay, fix is here, #21721

**Reply by Flolaffel** [reply]
Thanks, now `ComputeLinearElasticStress` works and delivers the same results as `ComputeFiniteStrainElasticStress`.

I have a follow-up question though: In you earlier comment you said to use the Green-Naghdi rate to match the results from the `StressDivergenceTensors`-System closely. But upon reading [Rashid (1993)](https://onlinelibrary.wiley.com/doi/abs/10.1002/nme.1620362302) that is referenced in your `ComputeFiniteStrain` documentation it seems to me that the `StressDivergenceTensors`-System should use the Jaumann rate. Am I wrong?

**Reply by reverendbedford** [reply]
My understanding is they advect with the rotation (R).  But I see the paper claims it's an implementation of the Jaumann rate.

I guess one solution is to run the same model with the "old" and "new" systems, using both rates for the "new" system and see which one is closest.
